### DISCUSSION TITLE
Brew doctor gives "Warning: Unknown support status"

### DISCUSSION THREAD

**Discussion by NathanSkene** [Homebrew/discussions#422] [Everyday usage]
Hi all,

I'm having some issues with brew so I'm trying to clear all `brew doctor` issues. I'm got a quite mysterious warning though.

"Warning: Unknown support status"

In [this thread](https://discourse.brew.sh/t/warning-unknown-support-status/8402/4) it was suggested that removing the command line tools and reinstalling might help. Tried that, didn't change anything.

Would really appreciate any further advice.

Output of git config below:

```
╰─ brew config                                                                                                                                                   ─╯
HOMEBREW_VERSION: 2.7.1
ORIGIN: https://github.com/Homebrew/brew
HEAD: fdb71a5cb3029da2ef74d7e09c40575bbfb4ec72
Last commit: 5 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 0e91ee3f4965b6848b3bd75b3342b365038b85c5
Core tap last commit: 2 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.FDVDOthtnJ/org.macports:0
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit skylake
Clang: 12.0 build 1200
Git: 2.30.0 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
Java: 12, 1.8.202.08, 1.8.0_201
macOS: 11.1-x86_64
CLT: 12.3.0.0.1.1607026830
Xcode: 12.3
XQuartz: 2.7.11 => /opt/X11
```

**Comment by carlocab** [comment]
Is `brew` giving you any problems other than the cryptic message? If it isn't, it's probably safe to ignore it.

If it is:

Can you post the full output of `brew doctor`?

Then do `brew update-reset` and see if you can still reproduce the error. But, before you do that, please note the following:
```
Usage: brew update-reset [repository]

Fetch and reset Homebrew and all tap repositories (or any specified
repository) using git(1) to their latest origin/master.

Note: this will destroy all your uncommitted or committed changes.
```

**Comment by NathanSkene** [comment]
Thanks for your reply! Here we go:

```
╰─ brew doctor                                                                                                                                                   ─╯
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Unknown support status
```

When I then run `brew update-reset` I get the following output:

```
╰─ brew update-reset                                                                                                                                             ─╯
==> Fetching /usr/local/Homebrew...
remote: Enumerating objects: 31, done.
remote: Counting objects: 100% (31/31), done.
remote: Total 50 (delta 30), reused 30 (delta 30), pack-reused 19
Unpacking objects: 100% (50/50), 11.27 KiB | 137.00 KiB/s, done.
From https://github.com/Homebrew/brew
   e5eb6a264..0648082d6  master     -> origin/master

==> Resetting /usr/local/Homebrew...
Branch 'master' set up to track remote branch 'master' from 'origin'.
Switched to and reset branch 'master'
Your branch is up to date with 'origin/master'.

==> Fetching /usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask...
remote: Enumerating objects: 11, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 12 (delta 7), reused 7 (delta 7), pack-reused 1
Unpacking objects: 100% (12/12), 105.55 KiB | 359.00 KiB/s, done.
From https://github.com/Homebrew/homebrew-cask
   5c984455b8..9f6459fad2  master     -> origin/master

==> Resetting /usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask...
Branch 'master' set up to track remote branch 'master' from 'origin'.
Reset branch 'master'
Your branch is up to date with 'origin/master'.

==> Fetching /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core...
remote: Enumerating objects: 14, done.
remote: Counting objects: 100% (14/14), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 16 (delta 12), reused 14 (delta 12), pack-reused 2
Unpacking objects: 100% (16/16), 21.11 KiB | 202.00 KiB/s, done.
From https://github.com/Homebrew/homebrew-core
   0e91ee3f49..8547fc672b  master     -> origin/master

==> Resetting /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core...
Branch 'master' set up to track remote branch 'master' from 'origin'.
Reset branch 'master'
Your branch is up to date with 'origin/master'.
```

`brew doctor` still gives the same output

**Comment by NathanSkene** [comment]
The reason I want to resolve the warning is that I get this error message when I try to do `brew upgrade`:

```
╰─ brew upgrade                                                                                                                                                  ─╯
==> Upgrading 1 outdated package:
ccache HEAD-1bb7f76 -> HEAD-1bb7f76
==> Upgrading ccache HEAD-1bb7f76 -> HEAD-1bb7f76
==> Cloning https://github.com/ccache/ccache.git
Updating /Users/natske/Library/Caches/Homebrew/ccache--git
==> Checking out branch master
Already on 'master'
Your branch is up to date with 'origin/master'.
HEAD is now at 61ce8c4 Only accept -f(no-)color-diagnostics for Clang
==> cmake .
==> make install
Last 15 lines from /Users/natske/Library/Logs/Homebrew/ccache/02.make:
[ 21%] Linking C static library libblake3.a
cd /tmp/ccache-20210102-31927-1qk7pj6/src/third_party/blake3 && /usr/local/Cellar/cmake/3.19.2/bin/cmake -P CMakeFiles/blake3.dir/cmake_clean_target.cmake
cd /tmp/ccache-20210102-31927-1qk7pj6/src/third_party/blake3 && /usr/local/Cellar/cmake/3.19.2/bin/cmake -E cmake_link_script CMakeFiles/blake3.dir/link.txt --verbose=1
/usr/bin/ar qc libblake3.a CMakeFiles/blake3.dir/blake3.c.o CMakeFiles/blake3.dir/blake3_dispatch_ccache.c.o CMakeFiles/blake3.dir/blake3_portable.c.o CMakeFiles/blake3.dir/blake3_sse2.c.o CMakeFiles/blake3.dir/blake3_sse41.c.o CMakeFiles/blake3.dir/blake3_avx2.c.o CMakeFiles/blake3.dir/blake3_avx512.c.o
/usr/bin/ranlib libblake3.a
[ 21%] Built target blake3
[ 22%] Linking CXX static library libthird_party_lib.a
cd /tmp/ccache-20210102-31927-1qk7pj6/src/third_party && /usr/local/Cellar/cmake/3.19.2/bin/cmake -P CMakeFiles/third_party_lib.dir/cmake_clean_target.cmake
cd /tmp/ccache-20210102-31927-1qk7pj6/src/third_party && /usr/local/Cellar/cmake/3.19.2/bin/cmake -E cmake_link_script CMakeFiles/third_party_lib.dir/link.txt --verbose=1
/usr/bin/ar qc libthird_party_lib.a CMakeFiles/third_party_lib.dir/base32hex.c.o CMakeFiles/third_party_lib.dir/format.cpp.o CMakeFiles/third_party_lib.dir/xxhash.c.o CMakeFiles/third_party_lib.dir/getopt_long.c.o CMakeFiles/third_party_lib.dir/xxh_x86dispatch.c.o
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libthird_party_lib.a(getopt_long.c.o) has no symbols
/usr/bin/ranlib libthird_party_lib.a
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libthird_party_lib.a(getopt_long.c.o) has no symbols
[ 22%] Built target third_party_lib
make: *** [all] Error 2

READ THIS: https://docs.brew.sh/Troubleshooting
```

**Reply by carlocab** [reply]
That's weird. `brew upgrade` should not be upgrading `HEAD` installs. Note that `HEAD` builds are also not supported. (i.e. I'll try to help you as and when I can, but we will always prioritise making sure `bottle` installs work first.)

Do you have MacPorts installed? That's known to not interact well with Homebrew installations.

**Reply by NathanSkene** [reply]
Thanks again for your reply! Afraid I don't know what you mean about `HEAD` vs `bottle` installs. Haven't got the faintest idea why one would occur versus the other. Don't think I've done anything unusual in my setup!

Didn't have macports installed on my zsh accessible part, but I used to use bash as my default shell and an old version was accessible from there, albeit it no longer worked. I've now uninstalled it. Didn't make any difference to either the `brew doctor` or `brew upgrade` errors.

**Reply by carlocab** [reply]
This line

    ==> Upgrading ccache HEAD-1bb7f76 -> HEAD-1bb7f76

means that `ccache` was installed using `brew install --HEAD ccache` (or perhaps reinstalled using those flags). I don't think the error is related to the warning you get from `brew doctor`. If you want to fix your `ccache` installation, I suggest using a non-`HEAD` install:
```
brew uninstall --force ccache
brew install ccache
```
You may need to do the `uninstall` step more than once in case you somehow have several versions of `ccache` installed for some reason.

Re Homebrew terminology: https://docs.brew.sh/Formula-Cookbook#homebrew-terminology

**Comment by Rylan12** [comment]
Can you paste the output of the following commands? It might help debug the issue. TBH I'm not super familiar with this code, though, so no promises 😅

(I agree, though, the upgrade issue may be unrelated. This error message is usually for casks not for formulae)

```sh
brew ruby -e 'require("cask");puts(Cask::Quarantine.check_quarantine_support)'
```

```sh
brew ruby -e 'puts DevelopmentTools.locate("swift")'
```

```sh
swift $(brew --repo)/Library/Homebrew/cask/utils/quarantine.swift; echo $?
```

**Comment by NathanSkene** [comment]
Thanks, that resolved the `brew upgrade` issue! I still get the `brew
doctor` warning though.

**Comment by NathanSkene** [comment]
Thanks for the suggestions Rylan!

```
╰─ brew ruby -e
'require("cask");puts(Cask::Quarantine.check_quarantine_support)'
unknown
```

The `swift` command gives a huge amount of bad looking output with error
messages... what is that one doing? Looks to me like that might be getting
us somewhere. I've pasted it all into a gist here:
https://gist.github.com/NathanSkene/5ee8a37ef603a937f90dad1e0af5c35e. The
top few lines are pasted below:

```
╰─ swift $(brew --repo)/Library/Homebrew/cask/utils/quarantine.swift; echo
$?                                 ─╯
/usr/local/include/module.modulemap:198:13: warning: umbrella directory
'xlocale' not found
   umbrella "xlocale"
            ^
/usr/local/include/module.modulemap:437:13: warning: umbrella directory
'_types' not found
   umbrella "_types"
            ^
/usr/local/include/module.modulemap:522:14: warning: umbrella directory
'sys/_types' not found
    umbrella "sys/_types"
             ^
/usr/local/include/module.modulemap:1230:12: warning: umbrella directory
'machine' not found
  umbrella "machine"
           ^
/usr/local/include/module.modulemap:1299:12: warning: umbrella directory
'hfs' not found
  umbrella "hfs"
           ^
/usr/local/include/module.modulemap:1349:13: warning: umbrella directory
'sys/_pthread' not found
   umbrella "sys/_pthread"
            ^
/usr/local/include/module.modulemap:20:10: error: header 'sys/cdefs.h' not
found
  header "sys/cdefs.h"
         ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib/swift/Darwin.swiftmodule/x86_64-apple-macos.swiftinterface:4:19:
error: could not build Objective-C module 'Darwin'
@_exported import Darwin
```

**Reply by Rylan12** [reply]
Okay, so that's definitely the issue. The `swift $(brew --repo)/Library/Homebrew/cask/utils/quarantine.swift` command should run successfully and exit with return code 2.

In terms of solving the issue, I'm not totally sure off the top of my head. I'll try to look into it and see if I can find any helpful links or anything.

---

In the meantime, I would double-check that there aren't any updates available for Xcode or the CLT. It may be worth checking `softwareupdate --all --list` for updates too.

If none are available, try reinstalling the CLT again. Make sure to fully remove the existing CLT. The location of the CLT can be found with `xcode-select -p`. For me, it's `/Library/Developer/CommandLineTools`. Then reinstall using either `xcode-select --install` or the latest stable version available from https://developer.apple.com/download/more/. I know you said you tried this, so I'm not really sure what to tell you.

Lastly, you can try uninstalling and reinstalling Xcode. That's probably a more significant task, so I'd leave it as a last resort.

To be honest, though, if isn't causing you a lot of pain, I'd just ignore it. If you don't use casks that often, you probably won't run into any issues. If you do try to install a cask and are having issues, try opening an issue over in the [Homebrew/homebrew-cask repo](https://github.com/Homebrew/homebrew-cask). The folks over there may be able to help you better than I can with specific cask issues.

Sorry I can't really be of more help here. I'll let you know if I find anything

**Comment by NathanSkene** [comment]
Hi Rylan,

Had actually reinstalled XCode and deleted+reinstalled CLI twice. Deleted
XCode from the Applications folder to enable reinstallation, but don't know
if there's a more rigourous way to remove it entirely.

This is stopping me from compiling any R packages... or at least digging
into this is what got me here. So it's a major problem.

Thanks!

**Comment by NathanSkene** [comment]
Hi Rylan,

Software update shows nothing.

```
╰─ softwareupdate --all --list
                                   ─╯
Software Update Tool

Finding available software
No new software available.
```

If it's any help, the error that got me digging into this is when I try to
run `pkgbuild::check_build_tools(debug = TRUE)` within R:

```
Trying to compile a simple C file
Running /Library/Frameworks/R.framework/Resources/bin/R CMD SHLIB foo.c
clang -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG
-I/usr/local/include   -fPIC  -Wall -g -O2  -c foo.c -o foo.o
clang -dynamiclib -Wl,-headerpad_max_install_names -undefined
dynamic_lookup -single_module -multiply_defined suppress
-L/Library/Frameworks/R.framework/Resources/lib -L/usr/local/lib -o foo.so
foo.o -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework
-Wl,CoreFoundation
ld: framework not found CoreFoundation
clang-7: error: linker command failed with exit code 1 (use -v to see
invocation)
make: *** [foo.so] Error 1
Error: Could not find tools necessary to compile a package
Call `pkgbuild::check_build_tools(debug = TRUE)` to diagnose the problem.
```

**Comment by NathanSkene** [comment]
A little update... don't know exactly how much these problems overlap but my guess is that they are related. I've solved the R package compilation problem by installing a [new version of gfortran](https://github.com/fxcoudert/gfortran-for-macOS/releases/tag/10.2-bigsur-intel).

I then had to set my Makevars file to point at a different version of gfortran to the one which I think is used by homebrew.

```
FLIBS=-L/usr/local/gfortran/lib/gcc/x86_64-apple-darwin15/6.1.0 -L/usr/local/gfortran/lib -lgfortran -lquadmath -lm
CXX1X=/usr/local/gfortran/bin/g++
CXX98=/usr/local/gfortran/bin/g++
CXX11=/usr/local/gfortran/bin/g++
CXX14=/usr/local/gfortran/bin/g++
CXX17=/usr/local/gfortran/bin/g++

LLVM_LOC = /usr/local/opt/llvm
CC=/usr/local/gfortran/bin/gcc -fopenmp
CXX=/usr/local/gfortran/bin/g++ -fopenmp
CFLAGS=-g -O3 -Wall -pedantic -std=gnu99 -mtune=native -pipe
CXXFLAGS=-g -O3 -Wall -pedantic -std=c++11 -mtune=native -pipe
LDFLAGS=-L/usr/local/opt/gettext/lib -L$(LLVM_LOC)/lib -Wl,-rpath,$(LLVM_LOC)/lib
CPPFLAGS=-I/usr/local/opt/gettext/include -I$(LLVM_LOC)/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
```

Note that differences in response to `gfortran -v` before and after installing the new version.... These make me think homebrew is implicated somehow.

Before install...

```
╰─ gfortran -v                                                                                         ─╯
Using built-in specs.
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/usr/local/Cellar/gcc/10.2.0_1/libexec/gcc/x86_64-apple-darwin20/10.2.0/lto-wrapper
Target: x86_64-apple-darwin20
Configured with: ../configure --build=x86_64-apple-darwin20 --prefix=/usr/local/Cellar/gcc/10.2.0_1 --libdir=/usr/local/Cellar/gcc/10.2.0_1/lib/gcc/10 --disable-nls --enable-checking=release --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-10 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-system-zlib --with-pkgversion='Homebrew GCC 10.2.0_1' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --disable-multilib --with-native-system-header-dir=/usr/include --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk SED=/usr/bin/sed
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 10.2.0 (Homebrew GCC 10.2.0_1)
```

After installing the software, this changed to:

```
╰─ gfortran -v                                                               ─╯
Using built-in specs.
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/usr/local/gfortran/libexec/gcc/x86_64-apple-darwin20/10.2.0/lto-wrapper
Target: x86_64-apple-darwin20
Configured with: ../gcc-10.2.0/configure --prefix=/usr/local/gfortran --with-gmp=/Users/fx/devel/gcc/build_package/deps --with-isl=/Users/fx/devel/gcc/build_package/deps --enable-languages=c,c++,fortran,objc,obj-c++ --build=x86_64-apple-darwin20 --with-native-system-header-dir=/usr/include --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 10.2.0 (GCC)
```

While I can now compile R packages (as R's makevars have been set to use the newly installed version), `brew doctor` still gives the same errors, as does the `swift` command.

**Reply by carlocab** [reply]
If you already have the `gcc` formula installed, then there's no need to install the version of `gfortran` that you linked to, as it is essentially identical to the `gcc` formula. (One difference: your `gfortran` is actually not compiled on Big Sur, so you may face some incompatibility issues.) You now have two copies of `gcc` on your system.

For details, see https://github.com/Homebrew/homebrew-cask/issues/96135.

To configure the `gcc` formula for use with R, see https://github.com/Homebrew/homebrew-cask/issues/96135#issuecomment-749549399 and succeeding comments.

**Comment by NathanSkene** [comment]
Hi Carlo,

Interesting, thanks.

I'd run `brew reinstall gcc` countless times before hand, and it didn't
resolve the issue. I guess it's the directory that they get installed in
that's the problem, and that the Makevars needs to be set appropriately as
in the link that you sent.

Unclear to me how I would uninstall the new version of gfortran, and as R
now works OK, unless it's breaking something else, or you think it's likely
to do so, I'm inclined to leave it well enough alone.

Would still be good to figure out what the "Warning: Unknown support
status" message is about though (and all the swift errors)! If they are in
some way related to the gfortran thing, then I'm happy to delete the new
version to resolve it.

Thanks again all!

**Reply by Rylan12** [reply]
> Warning: Unknown support status" message is about though (and all the swift errors)! If they are in some way related to the gfortran thing, then I'm happy to delete the new version to resolve it.

I don't think they're related.

Unfortunately, I haven't been able to find much to help here. Sorry about that. I did signal boost this to the rest of the maintainers so hopefully someone will chime in if they have any ideas.

I think it's fine to ignore the message if it's not causing any issues. If you _are_ having trouble installing something, your best bet is probably to open a GitHub issue and we can troubleshoot. Problems installing formulae should be reported to [Homebrew/homebrew-core](https://github.com/Homebrew/homebrew-core/) and problems with casks should be reported to [Homebrew/homebrew-cask](https://github.com/Homebrew/homebrew-cask/).
